CI: Improvement to speed up compilation and reduce download errors.

simbit18 opened the pull request:
## Summary

The simple improvement is designed to speed up compilation and reduce download errors on github and local.

Added a folder nxtmpdir for storing third-party packages

nuttxworkspace
         |
         |- nuttx
         |- apps
         |- nxtmpdir

tools/Unix.mk:
         added nxtmpdir folder creation

tools/configure.sh:
   added option -S creates the nxtmpdir folder for third-party packages.

For now I added in the folder this package

  ESP_HAL_3RDPARTY_URL = https://github.com/espressif/esp-hal-3rdparty.git

ARCH
  arch/xtensa/src/esp32/Make.defs
  arch/xtensa/src/esp32s2/Make.defs
  arch/xtensa/src/esp32s3/Make.defs
  arch/risc-v/src/common/espressif/Make.defs
  arch/risc-v/src/esp32c3-legacy/Make.defs

but you can also add other packages (maybe also of apps)


@no1wudi @tmedicci

## Impact
There should be no impact.
## Testing
CI

hartmannathan commented:
Do GitHub Actions download packages each time they run?

If so, can the packages be downloaded once and archived somehow, so they don't have to be re-downloaded every time the Action runs?

Is that what this PR is doing?

tmedicci commented:
Thanks @simbit18 . Let me test on our internal CI.

tmedicci commented on arch/risc-v/src/common/espressif/Make.defs:
```diff
@@ -146,7 +146,10 @@ endif
 chip/$(ESP_HAL_3RDPARTY_REPO):
 	$(Q) echo "Cloning Espressif HAL for 3rd Party Platforms"
-	$(Q) git clone --quiet $(ESP_HAL_3RDPARTY_URL) chip/$(ESP_HAL_3RDPARTY_REPO)
+	$(Q) if [ ! -d "$(NXTMPDIR)/$(ESP_HAL_3RDPARTY_REPO)" ]; then \
+		git clone --quiet $(ESP_HAL_3RDPARTY_URL) $(NXTMPDIR)/$(ESP_HAL_3RDPARTY_REPO); \
+	 fi
+	$(Q) cp -fr $(NXTMPDIR)/$(ESP_HAL_3RDPARTY_REPO) chip/$(ESP_HAL_3RDPARTY_REPO)
```
Let us combine the logic with the one in https://github.com/apache/nuttx/pull/13297. See: if we don't run `./tools/configure.sh -S` between two different NuttX versions, the build system will try to use the existing folder at the temp folder, but a different version may be required and this can be very dangerous (and hard to debug). So, we should an add an `else` to this clause to check if the version is the same.

tmedicci replied:
Or any equivalent logic: if the package exists and the version is the same, use it. Otherwise git clone it again.

simbit18 commented:
Hi @hartmannathan 

Let's use as an example architectures with expressif SoC.
( espressif developers i love you !!! :) )

For these jobs Linux (risc-v-01), Linux (xtensa-01), Linux (xtensa-02) and for <board name>:<board configuration> with the espressif/ESP-HAL-3RDPARTY
is cloned and then cleaned up each time even if it doesn't change until the job is finished.
So in this case but in general this logic slows down the compilation and can be more prone to network error resulting in workflow failure.

This is why I thought of a temporary storage folder.

In the case of compilation on GITHUB at each restart of the workflow for each job is:

1 created the temporary storage folder
2 cloned the repository the espressif/ESP-HAL-3RDPARTY only for the first time

the previous workflow does not change (logic of moving to branch-commit).


In the case of local compilation
1 the first compilation creates the temporary storage folder and clone espressif/ESP-HAL-3RDPARTY
2 the -S option removes the temporary folder with each new configuration .

the previous workflow does not change (logic of moving to branch-commit).


Rightly as @tmedicci  noted, since the third-party hal is not in the nuttx tree and we can't be aware of changes in the repository (logic of moving to branch-commit) users who don't update the configuration without the -S option might have an outdated repository locally.

Time comparison

Without this PR
![first](https://github.com/user-attachments/assets/f172fa50-618a-40ae-9cd9-09ceb1a73707)

https://github.com/apache/nuttx/actions/runs/10730602568/usage

With this PR
![after](https://github.com/user-attachments/assets/6d930fd4-2a12-4323-8101-50c93557ee23)
https://github.com/apache/nuttx/actions/runs/10722417518/usage

tmedicci commented:
> Hi @hartmannathan
> 
> Let's use as an example architectures with expressif SoC. ( espressif developers i love you !!! :) )
> 
> For these jobs **Linux (risc-v-01), Linux (xtensa-01), Linux (xtensa-02)** and for board name:board configuration with the espressif ESP-HAL-3RDPARTY is cloned and then cleaned up each time even if it doesn't change until the job is finished.
> 
> So in this case but in general this logic slows down the compilation and can be more prone to network error resulting in workflow failure.
> 
> This is why I thought of a temporary storage folder in this PR.
> 
> In the case of compiling on GITHUB at each restart of the workflow for each job it does these steps:
> 
> 1 created the temporary storage folder 2 clone the repository the espressif/ESP-HAL-3RDPARTY only for the first time
> 
> the previous workflow does not change (logic of moving to branch-commit).
> 
> In the case of local build 1 the first compiling creates the temporary storage folder and clone espressif/ESP-HAL-3RDPARTY 2 the -S option removes the temporary folder with each new configuration .
> 
> the previous workflow does not change (logic of moving to branch-commit).
> 
> Rightly as @tmedicci noted, since the espressif/ESP-HAL-3RDPARTY is not in the nuttx tree and we can't be aware of changes in the repository (logic of moving to branch-commit) users who don't update the configuration without the -S option might have an outdated repository locally.
> 
> Time comparison
> 
> Without this PR ![first](https://private-user-images.githubusercontent.com/101105604/365103853-f172fa50-618a-40ae-9cd9-09ceb1a73707.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2MjY3MDcsIm5iZiI6MTcyNTYyNjQwNywicGF0aCI6Ii8xMDExMDU2MDQvMzY1MTAzODUzLWYxNzJmYTUwLTYxOGEtNDBhZS05Y2Q5LTA5Y2ViMWE3MzcwNy5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwNlQxMjQwMDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYWMwNjFkYTlmZGUwMWQ3ZGMwNWJhZWRjMjQzZWQ2YTNiMGQ5MTgxYjQxZjg2NGUxNjUyNjExODA2MTVkMjk0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.uGG2ig22pnI2I0pL08Y3Pkrb1FfVrx0z2M3uV8_bSmw)
> 
> https://github.com/apache/nuttx/actions/runs/10730602568/usage
> 
> With this PR ![after](https://private-user-images.githubusercontent.com/101105604/365104514-6d930fd4-2a12-4323-8101-50c93557ee23.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU2MjY3MDcsIm5iZiI6MTcyNTYyNjQwNywicGF0aCI6Ii8xMDExMDU2MDQvMzY1MTA0NTE0LTZkOTMwZmQ0LTJhMTItNDMyMy04MTAxLTUwYzkzNTU3ZWUyMy5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwNlQxMjQwMDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZmMxNDMzNjBhOGE2ZWIzMDZmZDdiYTIwNDUwYTA4NTk5NTAxOGY1NTJlM2Q2Zjc5MGI2NTgzNDIxYWNiYjUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.KLVBf-5STQhT0ry1aev3sOhuo9V99vwdbvg3fFYqQtE) https://github.com/apache/nuttx/actions/runs/10722417518/usage

Yeah, this is an important improvement for our CI!

tmedicci commented:
> Rightly as @tmedicci noted, since the espressif/ESP-HAL-3RDPARTY is not in the nuttx tree and we can't be aware of changes in the repository (logic of moving to branch-commit) users who don't update the configuration without the -S option might have an outdated repository locally.

Just to make sure, do you agree about the problem of the not-in-sync repository? Are you planning to implement a workaround?

simbit18 commented:
> > Just to make sure, do you agree about the problem of the not-in-sync repository? Are you planning to implement a workaround?

HI @tmedicci  only locally  if user does not use ./tools/configure.sh -S esp32-devkitc:<config_name> could happen.

When I have time I will see if it is possible to find solution to correct this case.

Of course comments and suggestions are welcome !!! :)

xiaoxiang781216 commented on tools/Unix.mk:
```diff
@@ -19,6 +19,14 @@
+$(info $$NXTMPDIR = $(NXTMPDIR))
+
+ifeq ($(wildcard $(NXTMPDIR)),)
+$(shell mkdir -p $(NXTMPDIR))
```
but, it isn't good to always create a temp folder if esp isn't enable in defconfig.
BTW, it should be contained in ci specific script, Unix.mk it's too general.

simbit18 replied:
Hi @xiaoxiang781216  
my idea to speed up build is not only about esp but in general all parts where git and curl is used even to apps (draft for lvgl https://github.com/apache/nuttx-apps/commit/cbc3cc42985efcfcc2e999500498f557f282c9e3) when possible. 

.github/workflows/build.yml:
added option -S  adds the nxtmpdir folder for third-party packages.
./cibuild.sh -c -A -N -R -S testlist/${{matrix.boards}}.dat

simbit18 commented:
Update

changed logic -S now creates the nxtmpdir folder for third-party packages.

.github/workflows/build.yml:
added option -S adds the nxtmpdir folder for third-party packages.
./cibuild.sh -c -A -N -R -S testlist/${{matrix.boards}}.dat

@tmedicci you can run a test on your internal CI.

tmedicci commented:
> Updated
> 
> changed logic -S now creates the nxtmpdir folder for third-party packages. without -S the previous workflow does not change.
> 
> .github/workflows/build.yml: added option -S adds the nxtmpdir folder for third-party packages. ./cibuild.sh -c -A -N -R -S testlist/${{matrix.boards}}.dat
> 
> Hi @tmedicci you can run a test on your internal CI.

Hi @simbit18, thanks for updating it. Particularly, I liked the solution of creating the temp folder only if `-S` is passed for `./tools/configure.sh`. However, I'm still worried about the mismatch (if an user updates NuttX and the build system uses an outdated version of the HAL). But I think we can address it just like suggested by https://github.com/apache/nuttx/pull/13297 (@no1wudi ), what do you think? I think we can use check for the commit of the HAL in the temp folder and copy only if it is the same as the one we need.

xiaoxiang781216 commented on arch/risc-v/src/common/espressif/Make.defs:
```diff
@@ -146,7 +146,14 @@ endif
+				git clone --quiet $(ESP_HAL_3RDPARTY_URL) $(NXTMPDIR)/$(ESP_HAL_3RDPARTY_REPO); \
+			fi; \
+			cp -fr $(NXTMPDIR)/$(ESP_HAL_3RDPARTY_REPO) chip/$(ESP_HAL_3RDPARTY_REPO); \
+		 else \
```
should we add a download function to avoid the code dup? cmake version need similar change too.

simbit18 replied:
@xiaoxiang781216 Yes, I think it needs to be optimized and later for cmake as well.

simbit18 commented:
>> Hi @simbit18, thanks for updating it. Particularly, I liked the solution of creating the temp folder only if `-S` is passed for `./tools/configure.sh`. However, I'm still worried about the mismatch (if an user updates NuttX and the build system uses an outdated version of the HAL). But I think we can address it just like suggested by #13297 (@no1wudi ), what do you think? I think we can use check for the commit of the HAL in the temp folder and copy only if it is the same as the one we need.

Hi @tmedicci  I will also try to check the HAL commitment but I think to make it generic it has to be found in a script.

The idea of speeding up the NuttX build and using the tmp folder is not only about esp HAL but in general in all parts where git and curl are used.

Because surely the archs, boards and configs will increase in the future, and with them the time to finish jobs will also increase.

tmedicci commented:
> Hi @tmedicci I will also try to check the commit of the HAL but I think to make it generic it needs to be transferred to a shell scripting.

I agree! This is a starting point. From my perspective, it's totally acceptable to check the git version of the HAL in the makefile as the starting point ;)

simbit18 commented:
HI @tmedicci  @xiaoxiang781216 I simplified by adding these macros to config.mk.

```
# CLONE - Git clone repository. Initializes a new Git repository in the 
#         folder on your local machine and populates it with the contents
#         of the central repository.
#         The third argument is an storage path. The second argument is used
#         if it is not provided or is empty.
# Example: $(call CLONE,$(URL_BASE),$(PATH),$(STORAGE_FOLDER))

define CLONE
	$(ECHO_BEGIN)"Clone: $(if $3,$3,$2) "
	if [ -z $3 ]; then \
		git clone --quiet $1 $2; \
	else \
		if [ ! -d $3 ]; then \
			git clone --quiet $1 $3; \
		fi; \
		cp -fr $3 $2; \
	fi
	$(ECHO_END)
endef

# COMMITSHA - Check if the branche contain the named commit.
#         Remove the folder if the commit is not present in the branch.
#         The first argument is the repository folder on the local machine.
#         The second argument is unique SHA-1 hash value.
# Example: $(call COMMITSHA,$(GIT_FOLDER),$(COMMIT_SHA-1))

define COMMITSHA
	$(ECHO_BEGIN)"COMMIT SHA-1: $2 "
	if [ -d $1 ]; then \
		if ! git -C $1 branch --contains $2 > /dev/null 2>&1; then \
			rm -rf $1; \
		fi \
	fi
	$(ECHO_END)
endef

```
@tmedicci Now there is also version control.

tmedicci commented:
> HI @tmedicci @xiaoxiang781216 I simplified by adding these macros to config.mk.
> 
> ```
> # CLONE - Git clone repository. Initializes a new Git repository in the 
> #         folder on your local machine and populates it with the contents
> #         of the central repository.
> #         The third argument is an storage path. The second argument is used
> #         if it is not provided or is empty.
> # Example: $(call CLONE,$(URL_BASE),$(PATH),$(STORAGE_FOLDER))
> 
> define CLONE
> 	$(ECHO_BEGIN)"Clone: $(if $3,$3,$2) "
> 	if [ -z $3 ]; then \
> 		git clone --quiet $1 $2; \
> 	else \
> 		if [ ! -d $3 ]; then \
> 			git clone --quiet $1 $3; \
> 		fi; \
> 		cp -fr $3 $2; \
> 	fi
> 	$(ECHO_END)
> endef
> 
> # COMMITSHA - Check if the branche contain the named commit.
> #         Remove the folder if the commit is not present in the branch.
> #         The first argument is the repository folder on the local machine.
> #         The second argument is unique SHA-1 hash value.
> # Example: $(call COMMITSHA,$(GIT_FOLDER),$(COMMIT_SHA-1))
> 
> define COMMITSHA
> 	$(ECHO_BEGIN)"COMMIT SHA-1: $2 "
> 	if [ -d $1 ]; then \
> 		if ! git -C $1 branch --contains $2 > /dev/null 2>&1; then \
> 			rm -rf $1; \
> 		fi \
> 	fi
> 	$(ECHO_END)
> endef
> ```
> 
> @tmedicci Now there is also version control.

That's fine, thanks! I will test it in our internal CI

tmedicci commented on tools/configure.sh:
```diff
@@ -217,6 +223,19 @@ if [ -r ${dest_config} ]; then
+if [ "X${store_nxtmpdir}" = "Xy" ]; then
+  if [ ! -d "${WSDIR}/nxtmpdir" ]; then
+    mkdir -p "${WSDIR}/nxtmpdir"
+    echo "Folder nxtmpdir create."
```
Should we print the whole path for the folder?

tmedicci replied:
typo: `created`.

tmedicci commented on tools/Config.mk:
```diff
@@ -608,6 +620,42 @@ define DOWNLOAD
+	$(ECHO_END)
+endef
+
+# COMMITSHA - Check if the branche contain the named commit.
+#         Remove the folder if the commit is not present in the branch.
+#         The first argument is the repository folder on the local machine.
+#         The second argument is unique SHA-1 hash value.
+# Example: $(call COMMITSHA,$(GIT_FOLDER),$(COMMIT_SHA-1))
```
I would call it `CHECK_COMMITSHA`. Also, I think the feature of this function should be different: instead of using `git branch --contains`, I think we should use `git rev-parse HEAD` to get the latest commit's SHA-1 and compare it with the provided one. Note: the branch may contain the commit but may be at a newer HEAD and, then, we wouldn't be using the same commit as expected. To solve that we should stick with the HEAD only.

tmedicci replied:
Forget about it: if it contains the commit, we would checkout to it soon. So, just some minor changes:
```suggestion
# CHECK_COMMITSHA - Check if the branch contains the commit SHA-1.
#         Remove the folder if the commit is not present in the branch.
#         The first argument is the repository folder on the local machine.
#         The second argument is a unique SHA-1 hash value.
# Example: $(call CHECK_COMMITSHA,$(GIT_FOLDER),$(COMMIT_SHA-1))
```

tmedicci replied:
Also, I think we can add more verbosity when the folder is non-existing and when it exists but contains no reference for the commit.

tmedicci commented on tools/Config.mk:
```diff
@@ -608,6 +620,42 @@ define DOWNLOAD
+
+define CLONE
+	$(ECHO_BEGIN)"Clone: $(if $3,$3,$2) "
+	if [ -z $3 ]; then \
+		git clone --quiet $1 $2; \
+	else \
+		if [ ! -d $3 ]; then \
+			git clone --quiet $1 $3; \
+		fi; \
+		cp -fr $3 $2; \
+	fi
```
Same here: we can add more verbosity for the cases where we provide the storage folder and copy from the storage folder to the final path.

xiaoxiang781216 commented on arch/risc-v/src/common/espressif/Make.defs:
```diff
@@ -146,7 +148,12 @@ endif
 chip/$(ESP_HAL_3RDPARTY_REPO):
 	$(Q) echo "Cloning Espressif HAL for 3rd Party Platforms"
-	$(Q) git clone --quiet $(ESP_HAL_3RDPARTY_URL) chip/$(ESP_HAL_3RDPARTY_REPO)
+ifeq ($(STORAGETMP),y)
```
could we simply call one function with the enough arguments, which could do all thing internally?

simbit18 commented:
HI @tmedicci @xiaoxiang781216 I updated the commit and also added arch/risc-v/src/esp32c3-legacy/Make.defs

tmedicci reviewed (approved):
LGTM!

jerpelea reviewed (approved):
please add commit messages

simbit18 commented:
====================================================================================
Configuration/Tool: stm32f4discovery/nsh,CONFIG_ARM_TOOLCHAIN_GNU_EABI
2024-09-17 07:47:16
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
make: *** [olddefconfig] Error 1
  Enabling CONFIG_ARM_TOOLCHAIN_GNU_EABI
make: *** [olddefconfig] Error 1
  Building NuttX...
make[1]: *** No rule to make target `context'.

It does not depend on this PR.
I will rebase to see if this can fix the CI error.

simbit18 commented:
> please add commit messages

Hi @jerpelea after passing the CI I will squash the commits into a single commit with message.

simbit18 commented:
@jerpelea done
